src/ImageMetadata.jl

```diff
@@ -32,29 +33,38 @@ export
 Construct an image with `ImageMeta(A, props)` (for a properties dictionary
 `props`), or with `ImageMeta(A, prop1=val1, prop2=val2, ...)`.
 """
-mutable struct ImageMeta{T,N,A<:AbstractArray,P<:AbstractDict{String,Any}} <: AbstractArray{T,N}
+mutable struct ImageMeta{T,N,A<:AbstractArray,P<:AbstractDict{Symbol,Any}} <: AbstractArray{T,N}
```

Comment:
Does anything break if we change this to just plain `struct`? I ask because it might yield performance improvements for certain operations.

Reply:
With `mutable`
```julia
img = ImageMeta(rand(4,4))

@btime setproperty!(img, :prop1, [1,2,3])

  48.967 ns (1 allocation: 112 bytes)
Dict{Symbol,Any} with 1 entry:
  :prop1 => [1, 2, 3]

@btime getproperty(img, :prop1)
  22.753 ns (0 allocations: 0 bytes)
3-element Array{Int64,1}:
 1
 2
 3
```
Without
```julia
img = ImageMeta(rand(4,4))

julia> @btime setproperty!(img, :prop1, [1,2,3])
  48.609 ns (1 allocation: 112 bytes)
Dict{Symbol,Any} with 1 entry:
  :prop1 => [1, 2, 3]

julia> @btime getproperty(img, :prop1)
  22.099 ns (0 allocations: 0 bytes)
3-element Array{Int64,1}:
 1
 2
 3
```

Reply:
Dict accesses are so slow that I very much doubt you'd see any effect there. Array element accesses might be faster with `struct` (immutability is sometimes a huge advantage for compiler optimizations). 

In general I'm not sure I see a reason to allow people to swap in a different data array or different properties dict and call it the same object; the point of the wrapper seems to be to associate *this* properties dict with *this* array. But I am willing to be persuaded otherwise.

Reply:
> In general I'm not sure I see a reason to allow people to swap in a different data array or different properties dict and call it the same object

I completely agree with this sentiment. I should have clarified what I was demonstrating with the benchmark. It shows a pretty small improvement, but I'm pretty sure that it could result in some nice performance gains when used in combination with some custom properties fields.

Reply:
Good. Other benchmarks:
```julia
julia> struct WrapperI{A,P}
           data::A
           props::P
       end

julia> mutable struct WrapperM{A,P}
           data::A
           props::P
       end

julia> Base.getindex(w::WrapperI, i) = w.data[i]

julia> Base.getindex(w::WrapperM, i) = w.data[i]

julia> data = rand(100); props = Dict(:hi=>"everyone")
Dict{Symbol,String} with 1 entry:
  :hi => "everyone"

julia> wi = WrapperI(data, props);

julia> wm = WrapperM(data, props);

julia> using BenchmarkTools

julia> @btime $wi[3]
  1.481 ns (0 allocations: 0 bytes)
0.4001678030037503

julia> @btime $wm[3]
  1.973 ns (0 allocations: 0 bytes)
0.4001678030037503
```
So it's a fairly large ratio.

Reply:
This is implemented now.  But there's a bit of quirky behavior when trying to set the `data` property.
```julia
ulia> d1 = rand(2,2)
2×2 Array{Float64,2}:
 0.32082   0.533365
 0.719087  0.432655

julia> d2 = rand(2,2)
2×2 Array{Float64,2}:
 0.491014  0.665869
 0.502986  0.535114

julia> img = ImageMeta(d1)
Float64 ImageMeta with:
  data: 2×2 Array{Float64,2}
  properties:

julia> img.data = d2
2×2 Array{Float64,2}:
 0.491014  0.665869
 0.502986  0.535114

julia> img
Float64 ImageMeta with:
  data: 2×2 Array{Float64,2}
  properties:
    data: [0.491014 0.665869; 0.502986 0.535114]
```

It's not completely clear to me what the best way to handle this is. I could do something like this

```julia
function Base.setproperty!(img::ImageMeta, propname::Symbol, val)
    # TODO remove these once deprecations are done
    if propname === :data
        Base.depwarn("setproperty(img, data, val) is deprecated use setindex!(img, :, val) instead", :rawdata)
        setindex!(img, val, img)
    elseif propname === :properties
        Base.depwarn("img.properties is deprecated use properties(img) to extract data img.", :properties)
        for (k,v) in val
            setproperty!(img, k, v)
        end
    else
        properties(img)[propname] = val
    end
    return img
end
```

Reply:
Do we have to support setting data? For example, is there a test that fails? What I want to move to is that if you want to replace the data, just create a new `ImageMeta` wrapper.

Reply:
> Do we have to support setting data?

No. I'm just trying to cover all our bases for deprecation warnings.

> For example, is there a test that fails? What I want to move to is that if you want to replace the data, just create a new ImageMeta wrapper.

No test fails. Just trying to be thorough here. Because if someone was using `img.data = rand(2,2)` before this PR they might be upset that it just sets a property named `data` now. I don't know why anyone would be replacing these fields so if everyone's fine with the way `setproperty!` works right now then I'll just leave it.

Reply:
OK. Yes, let's include the functionality and deprecation warning, but also mention creating a new wrapper as an option.

Reply:
`setproperty!` deprecations is now implemented with suggestions to make new `ImageMeta` instead.